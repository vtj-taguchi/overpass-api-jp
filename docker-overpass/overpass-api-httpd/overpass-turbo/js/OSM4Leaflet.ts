import osmtogeojson from "osmtogeojson";

type Options = {
  data_mode: string;
  baseLayerClass: typeof L.GeoJSON;
  baseLayerOptions: L.GeoJSONOptions;
  afterParse?: () => void;
};

class OSM4Leaflet extends L.Layer {
  private _resultData: GeoJSON.FeatureCollection;
  private _baseLayer: L.GeoJSON;
  private options: Options = {
    data_mode: "xml",
    baseLayerClass: L.GeoJSON,
    baseLayerOptions: {}
  };

  constructor(data, options: Partial<Options>) {
    super();
    L.Util.setOptions(this, options);

    this._baseLayer = new this.options.baseLayerClass(
      null,
      this.options.baseLayerOptions
    );
    this._resultData = null;
    // if data
    if (data) this.addData(data);
  }
  addData(data, onDone?: () => void) {
    setTimeout(() => {
      // 1. convert to GeoJSON
      const geojson = osmtogeojson(data, {flatProperties: false});
      this._resultData = geojson;
      if (this.options.afterParse) this.options.afterParse();
      setTimeout(() => {
        // 2. add to baseLayer
        this._baseLayer.addData(geojson);
        if (onDone) onDone();
      }, 1); //end setTimeout
    }, 1); //end setTimeout
  }
  getGeoJSON() {
    return this._resultData;
  }
  getBaseLayer() {
    return this._baseLayer;
  }
  onAdd(map) {
    this._baseLayer.addTo(map);
    return this;
  }
  onRemove(map) {
    map.removeLayer(this._baseLayer);
    return this;
  }
}

export default OSM4Leaflet;
