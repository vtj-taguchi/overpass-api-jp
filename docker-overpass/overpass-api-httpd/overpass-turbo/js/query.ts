// query parser module
import _ from "lodash";
import type {Shortcut} from "./shortcuts";

export default class parser {
  statements: Record<string, string>;
  async parse(
    query: string,
    shortcuts: Record<string, Shortcut>,
    _found_statements?: Record<string, string>
  ): Promise<string> {
    // 1. get user defined constants
    const constants = {};
    const constant = /{{([A-Za-z0-9_]+)=(.+?)}}/;
    let c;
    while ((c = query.match(constant))) {
      const c_name = c[1];
      const c_val = c[2];
      constants[c_name] = c_val;
      // remove constant definitions
      query = query.replace(constant, "");
    }
    _.extend(shortcuts, constants, (b, a) => (typeof a == "undefined" ? b : a));
    // 2. replace overpass turbo this.statements, user-constants and shortcuts
    this.statements = {};
    if (_found_statements) this.statements = _found_statements;
    const statement = /{{([A-Za-z0-9_]+)(:([\s\S]*?))?}}/;
    let s;
    while ((s = query.match(statement))) {
      const s_name = s[1];
      const s_instr = s[3] || "";
      let s_replace = "";
      // save instructions for later
      if (this.statements[s_name] === undefined) this.statements[s_name] = "";
      this.statements[s_name] += s_instr;
      // if the statement is a shortcut, replace its content
      const shortcut = shortcuts[s_name];
      if (shortcut !== undefined) {
        // these shortcuts can also be callback functions, like {{date:-1day}}
        if (typeof shortcut === "function") {
          const res = await new Promise<string>((resolve) =>
            shortcut(s_instr, (s) => resolve(s))
          );
          const seed = Math.round(Math.random() * Math.pow(2, 22)); // todo: use some kind of checksum of s_instr if possible
          shortcuts[`__statement__${s_name}__${seed}`] = res;
          query = query.replace(
            s[0],
            `{{__statement__${s_name}__${seed}:${s_instr}}}`
          );
          // recursively call the parser with updated shortcuts
          return this.parse(query, shortcuts, this.statements);
        } else s_replace = shortcut;
      }
      // remove statement, but preserve number of newlines
      const lc = s_instr.split(/\r?\n|\r/).length;
      query = query.replace(s[0], s_replace + Array(lc).join("\n"));
    }
    // 3. remove remaining (e.g. unknown) mustache templates:
    let m;
    while ((m = query.match(/{{[\S\s]*?}}/gm))) {
      // count lines in template and replace mustache with same number of newlines
      const lc = m[0].split(/\r?\n|\r/).length;
      query = query.replace(m[0], Array(lc).join("\n"));
    }
    // return the query
    return query;
  }
  hasStatement(statement) {
    // eslint-disable-next-line no-prototype-builtins
    return this.statements.hasOwnProperty(statement);
  }
  getStatement(statement) {
    return this.statements[statement];
  }
}
