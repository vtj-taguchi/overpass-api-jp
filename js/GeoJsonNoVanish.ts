class GeoJsonNoVanish extends L.GeoJSON {
  threshold = 10;
  constructor(geojson, options) {
    super(geojson, options);
    this.threshold = options?.threshold ?? 10;
  }
  onAdd(map) {
    this._map = map;
    this.eachLayer(map.addLayer, map);

    this._map.addEventListener("zoomend", this._onZoomEnd, this);
    this._onZoomEnd();
    return this;
  }
  onRemove(map) {
    this._map.removeEventListener("zoomend", this._onZoomEnd, this);

    this.eachLayer(map.removeLayer, map);
    this._map = null;
    return this;
  }
  _onZoomEnd() {
    // todo: name
    // todo: possible optimizations: zoomOut = skip already compressed objects (and vice versa)
    const is_max_zoom = this._map.getZoom() == this._map.getMaxZoom();
    this.eachLayer(function (o) {
      if (!o.feature || !o.feature.geometry) return; // skip invalid layers
      if (o.feature.geometry.type == "Point" && !o.obj) return; // skip node features
      const compress =
        this.options.compress &&
        this.options.compress(o.obj ? o.obj.feature : o.feature);
      const crs = this._map.options.crs;
      if (o.obj) {
        if (compress === "point") return;
        // already compressed feature
        const bounds = o.obj.getBounds();
        const p1 = crs.latLngToPoint(bounds.getSouthWest(), o._map.getZoom());
        const p2 = crs.latLngToPoint(bounds.getNorthEast(), o._map.getZoom());
        const d = Math.pow(p1.x - p2.x, 2) + Math.pow(p1.y - p2.y, 2);
        if (d > Math.pow(this.threshold, 2) || is_max_zoom) {
          delete o.obj.placeholder;
          this.removeLayer(o);
          if (o._tooltip) {
            o.obj.bindTooltip(o._tooltip);
          }
          this.addLayer(o.obj);
        }
        return;
      }
      if (is_max_zoom && compress !== "point") return; // do not compress objects at max zoom, except if mapcss says always to render as points
      if (compress === "native") return; // do not compress if mapcss specifies not to
      const bounds = o.getBounds();
      const p1 = crs.latLngToPoint(bounds.getSouthWest(), o._map.getZoom());
      const p2 = crs.latLngToPoint(bounds.getNorthEast(), o._map.getZoom());
      const d = Math.pow(p1.x - p2.x, 2) + Math.pow(p1.y - p2.y, 2);
      if (d > Math.pow(this.threshold, 2) && compress !== "point") return;
      let center;
      if (d <= Math.pow(this.threshold, 2)) {
        center = bounds.getCenter();
      } else {
        center = o.getCenter();
      }
      const f = L.extend({}, o.feature);
      f.is_placeholder = true;
      f.geometry = {
        type: "Point",
        coordinates: [center.lng, center.lat]
      };
      const c = L.GeoJSON.geometryToLayer(f, this.options);
      o.placeholder = c;
      c.feature = f;
      c.obj = o;
      c.on("click", function (e) {
        this.obj.fireEvent(e.type, e);
      });
      this.removeLayer(o);
      this.resetStyle(c);
      c.options.interactive = true;
      c.options.stroke = true;
      c.options.fill = true;
      if (o._tooltip) {
        c.bindTooltip(o._tooltip);
      }
      this.addLayer(c);
    }, this);
  }
}

export default GeoJsonNoVanish;
