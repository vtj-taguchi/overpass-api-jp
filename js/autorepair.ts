// query autorepair module
import $ from "jquery";

import {Base64} from "./misc";

export default function autorepair(q, lng) {
  const repair = {};

  const comments = {};

  (function init() {
    // replace comments with placeholders
    // (we do not want to autorepair stuff which is commented out.)
    let cs, placeholder;
    if (lng == "xml") {
      cs = q.match(/<!--[\s\S]*?-->/g) || [];
      for (const csi of cs) {
        placeholder = `<!--${Base64.encode(Math.random().toString())}-->`; //todo: use some kind of checksum or hash maybe?
        q = q.replace(csi, placeholder);
        comments[placeholder] = csi;
      }
    } else {
      cs = q.match(/\/\*[\s\S]*?\*\//g) || []; // multiline comments: /*...*/
      for (const csi of cs) {
        placeholder = `/*${Base64.encode(Math.random().toString())}*/`; //todo: use some kind of checksum or hash maybe?
        q = q.replace(csi, placeholder);
        comments[placeholder] = csi;
      }
      cs = q.match(/\/\/[^\n]*/g) || []; // single line comments: //...
      for (const csi of cs) {
        placeholder = `/*${Base64.encode(Math.random().toString())}*/`; //todo: use some kind of checksum or hash maybe?
        q = q.replace(csi, placeholder);
        comments[placeholder] = csi;
      }
    }
  })();

  repair.getQuery = () => {
    // expand placeholded comments
    for (const placeholder in comments) {
      q = q.replace(placeholder, comments[placeholder]);
    }
    return q;
  };

  repair.recurse = () => {
    if (lng == "xml") {
      // do some fancy mixture between regex magic and xml as html parsing :â‚¬
      const prints =
        q.match(/(\n?[^\S\n]*<print[\s\S]*?(\/>|<\/print>))/g) || [];
      for (let i = 0; i < prints.length; i++) {
        const ws = prints[i].match(/^\n?(\s*)/)[1]; // amount of whitespace in front of each print statement
        const from = $("print", $.parseXML(prints[i])).attr("from");
        let add1, add2, add3;
        if (from) {
          add1 = ` into="${from}"`;
          add2 = ` set="${from}"`;
          add3 = ` from="${from}"`;
        } else {
          add1 = "";
          add2 = "";
          add3 = "";
        }
        q = q.replace(
          prints[i],
          `\n${ws}<!-- added by auto repair -->\n${ws}<union${add1}>\n${ws}  <item${add2}/>\n${ws}  <recurse${add3} type="down"/>\n${ws}</union>\n${ws}<!-- end of auto repair --><autorepair>${i}</autorepair>`
        );
      }
      for (let i = 0; i < prints.length; i++)
        q = q.replace(`<autorepair>${i}</autorepair>`, prints[i]);
    } else {
      const outs = q.match(/(\n?[^\S\n]*(\.[^.;]+)?out[^:;"\]]*;)/g) || [];
      for (let i = 0; i < outs.length; i++) {
        const ws = outs[i].match(/^\n?(\s*)/)[0]; // amount of whitespace
        const from = outs[i].match(/\.([^;.]+?)\s+out/);
        let add;
        if (from) add = `(.${from[1]};.${from[1]} >;)->.${from[1]};`;
        else add = "(._;>;);";
        q = q.replace(
          outs[i],
          `${ws}/*added by auto repair*/${ws}${add}${ws}/*end of auto repair*/<autorepair>${i}</autorepair>`
        );
      }
      for (let i = 0; i < outs.length; i++)
        q = q.replace(`<autorepair>${i}</autorepair>`, outs[i]);
    }
    return true;
  };

  repair.editors = () => {
    if (lng == "xml") {
      // 1. fix <osm-script output=*
      const src = q.match(/<osm-script([^>]*)>/);
      if (src) {
        const output = $(
          "osm-script",
          $.parseXML(`${src[0]}</osm-script>`)
        ).attr("output");
        if (output && output != "xml") {
          const new_src = src[0].replace(output, "xml");
          q = q.replace(src[0], `${new_src}<!-- fixed by auto repair -->`);
        }
      }
      // 2. fix <print mode=*
      const prints = q.match(/(<print[\s\S]*?(\/>|<\/print>))/g) || [];
      for (const print_i of prints) {
        const print = $("print", $.parseXML(print_i)),
          mode = print.attr("mode"),
          geometry = print.attr("geometry");
        let add = "";
        let new_print;
        let repaired = false;
        if (mode !== "meta") {
          print.attr("mode", "meta");
          repaired = true;
        }
        if (geometry && geometry !== "skeleton") {
          print.attr("geometry", null);
          const out_set = print.attr("from");
          if (!out_set) {
            add = '<union><item/><recurse type="down"/></union>';
          } else {
            add = `<union into="${out_set}"><item set="${out_set}"/><recurse from="${out_set}" type="down"/></union>`;
          }
          repaired = true;
        }
        if (repaired) {
          new_print = add + new XMLSerializer().serializeToString(print[0]);
          new_print += "<!-- fixed by auto repair -->";
          q = q.replace(print_i, new_print);
        }
      }
    } else {
      // 1. fix [out:*]
      const out = q.match(/\[\s*out\s*:\s*([^\]\s]+)\s*\]\s*;?/);
      ///^\s*\[\s*out\s*:\s*([^\]\s]+)/);
      if (out && out[1] != "xml")
        q = q.replace(
          /(\[\s*out\s*:\s*)([^\]\s]+)(\s*\]\s*;?)/,
          "$1xml$3/*fixed by auto repair*/"
        );
      // 2. fix print statements: non meta output, overpass geometries
      const prints = q.match(/(\.([^;.]+?)\s+)?(out[^:;"\]]*;)/g) || [];
      for (const print_i of prints) {
        // eslint-disable-next-line prefer-const
        let [print, , out_set, out_statement] = print_i.match(
          /(\.([^;.]+?)\s+)?(out[^:;"\]]*;)/
        );
        let new_print = print;
        let new_out_statement;
        // non meta output
        if (
          out_statement.match(/\s(body|skel|ids|tags|count)/) ||
          !out_statement.match(/\s(meta)/)
        ) {
          new_out_statement = out_statement
            .replace(/\s(body|skel|ids|tags|count|meta)/g, "")
            .replace(/^out/, "out meta");
          new_print = new_print.replace(out_statement, new_out_statement);
          out_statement = new_out_statement;
        }
        // overpass geometry modes
        if (out_statement.match(/\s(center|bb|geom)/)) {
          new_out_statement = out_statement.replace(/\s(center|bb|geom)/g, "");
          new_print = new_print.replace(out_statement, new_out_statement);
          out_statement = new_out_statement;
          if (out_set) {
            new_print = `(.${out_set};.${out_set} >;)->.${out_set}; ${new_print}`;
          } else {
            new_print = `(._;>;); ${new_print}`;
          }
        }
        if (new_print != print)
          q = q.replace(print, `${new_print}/*fixed by auto repair*/`);
      }
    }
    return true;
  };

  return repair;
}

autorepair.detect = {};
autorepair.detect.editors = (q, lng) => {
  // todo: test this
  // todo: move into autorepair "module" /// todo. done?
  q = q.replace(/{{.*?}}/g, "");
  const err = {};
  if (lng == "xml") {
    try {
      const xml = $.parseXML(`<x>${q}</x>`);
      const out = $("osm-script", xml).attr("output");
      if (out !== undefined && out !== "xml") err.output = true;
      $("print", xml).each((i, p) => {
        if ($(p).attr("mode") !== "meta") err.meta = true;
      });
      $("print", xml).each((i, p) => {
        if (
          $(p)
            .attr("geometry")
            .match(/(center|bounds|full)/)
        )
          err.geometry = true;
      });
    } catch (e) {} // ignore xml syntax errors ?!
  } else {
    // ignore comments
    q = q.replace(/\/\*[\s\S]*?\*\//g, "");
    q = q.replace(/\/\/[^\n]*/g, "");
    const out = q.match(/\[\s*out\s*:\s*([^\]\s]+)\s*\]/);
    if (out && out[1] != "xml") err.output = true;
    const prints = q.match(/out([^:;]*);/g);
    $(prints).each((i, p) => {
      if (p.match(/\s(body|skel|ids|tags)/) || !p.match(/meta/))
        err.meta = true;
    });
    $(prints).each((i, p) => {
      if (p.match(/\s(center|bb|geom)/)) err.geometry = true;
    });
  }
  return $.isEmptyObject(err);
};
