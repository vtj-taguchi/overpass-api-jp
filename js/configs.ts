export default {
  appname: "overpass-turbo",
  // used for localStorage and openstreetmap.org/api/0.6/user/preferences
  settingNamespace: "overpass-ide",
  defaultServer: "https://overpass-api.de/api/",
  // https://wiki.openstreetmap.org/wiki/Overpass_API#Public_Overpass_API_instances
  suggestedServers: [
    "https://overpass-api.de/api/",
    "https://overpass-api.de/api/",
    "https://maps.mail.ru/osm/tools/overpass/api/",
    "https://overpass.openstreetmap.ru/api/",
    "https://overpass.kumi.systems/api/"
  ],
  defaultTiles: "https://tile.openstreetmap.org/{z}/{x}/{y}.png",
  // https://wiki.osmfoundation.org/wiki/Licence/Attribution_Guidelines
  // > Attribution must be to "OpenStreetMap".
  // > Attribution must also make it clear that the data is available under the Open Database License.
  // > This may be done by making the text "OpenStreetMap" a link to openstreetmap.org/copyright, which has information about OpenStreetMap’s data sources (which OpenStreetMap needs to credit) as well as the ODbL.
  // https://www.openstreetmap.org/copyright
  tileServerAttribution: `© <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors`,
  suggestedTiles: [
    "https://tile.openstreetmap.org/{z}/{x}/{y}.png"
    //"http://{s}.tile.opencyclemap.org/cycle/{z}/{x}/{y}.png",
    //"http://{s}.tile2.opencyclemap.org/transport/{z}/{x}/{y}.png",
    //"http://{s}.tile3.opencyclemap.org/landscape/{z}/{x}/{y}.png",
    //"http://otile1.mqcdn.com/tiles/1.0.0/osm/{z}/{x}/{y}.jpg",
  ],
  defaultMapView: {
    lat: 41.89,
    lon: 12.492,
    zoom: 16
  },
  maxMapZoom: 20,
  short_url_service: "",
  push_history_url: true,
  html2canvas_use_proxy: false,
  // api key for osmnames geocoder, go to http://osmnames.org/api/ to get one if you run your own overpass instance
  osmnamesApiKey: "gtXyh2mBSaN5zWqqqQRh",
  // osmAuth configuration object (used for syncing saved queries). expects an osm-auth config object (min. the oauth_consumer_key and oauth_secret must be given), see https://github.com/osmlab/osm-auth#getting-keys
  osmAuth: {
    url: "https://www.openstreetmap.org",
    client_id: "lIifli2M7Enpi1LUqCxSNe3yDXhBHwf_n8HzJ03mKFg"
  }
};
