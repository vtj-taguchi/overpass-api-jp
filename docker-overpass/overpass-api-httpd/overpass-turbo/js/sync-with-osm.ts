import {osmAuth} from "osm-auth";
import configs from "./configs";

const enabled = configs.osmAuth && configs.osmAuth.client_id;

if (!configs.osmAuth.redirect_uri) {
  configs.osmAuth.redirect_uri = `${window.location.origin}${window.location.pathname}land.html`;
}
configs.osmAuth.scope = "read_prefs write_prefs";

let auth;
if (enabled) auth = osmAuth(configs.osmAuth);

export default {
  enabled: enabled,
  load(callback) {
    if (!auth.authenticated()) {
      auth.authenticate((err) => {
        //check err param exists
        if (err) return callback(err);
        loadQueries(callback);
      });
    } else {
      loadQueries(callback);
    }
  },
  save(query, callback) {
    if (!auth.authenticated()) {
      auth.authenticate((err) => {
        //check err param exists
        if (err) return callback(err);
        saveQuery(query, callback);
      });
    } else {
      saveQuery(query, callback);
    }
  },
  delete(query, callback) {
    if (!auth.authenticated())
      return callback(new Error("must be logged in to delete a synced query"));
    query = {
      name: query,
      deleteMe: true
    };
    saveQuery(query, callback);
  },
  logout() {
    if (auth.authenticated()) auth.logout();
  },
  authenticated() {
    return enabled && auth.authenticated();
  }
};

function loadQueries(callback) {
  auth.xhr(
    {
      method: "GET",
      path: "/api/0.6/user/preferences"
    },
    (err, res) => {
      if (err) return callback(err);

      let pref_count = 0,
        cnt_elem;
      if (
        (cnt_elem = res.querySelector(
          `preference[k="${configs.settingNamespace}_query-count"]`
        ))
      )
        pref_count = +cnt_elem.getAttribute("v");

      const result = [];
      for (let i = 0; i < pref_count; i++) {
        const pref_elem = res.querySelector(
          `preference[k="${configs.settingNamespace}_query_${i}_0"]`
        );
        if (!pref_elem) continue;
        const first_chunk = pref_elem.getAttribute("v").split("&");
        const length = +first_chunk[0].slice(2);
        const name = first_chunk[1].slice(2);
        let query = first_chunk[2].slice(2);
        for (let j = 1; j < length; j++) {
          query += res
            .querySelector(
              `preference[k="${configs.settingNamespace}_query_${i}_${j}"]`
            )
            .getAttribute("v");
        }
        result.push({
          name: name,
          query: query
        });
      }

      callback(null, result, res);
    }
  );
}

function saveQuery(new_query, callback) {
  loadQueries((err, existing_queries, dom: Document) => {
    if (err) return callback(err);

    const preferences = dom.querySelector("preferences");
    // clean up existing data
    const existing = preferences.querySelectorAll(
      `preference[k^="${configs.settingNamespace}_query"]`
    );
    for (const e of existing) {
      e.remove();
    }
    // insert new query into list of existing ones
    let is_new = true;
    existing_queries.forEach((q, idx) => {
      if (q.name == new_query.name) {
        q.query = new_query.query;
        is_new = idx;
      }
    });
    if (is_new === true) {
      existing_queries.push(new_query);
    } else if (new_query.deleteMe) {
      existing_queries.splice(is_new, 1);
    }
    // construct new preferences xml
    const new_elem = dom.createElement("preference");
    new_elem.setAttribute("k", `${configs.settingNamespace}_query-count`);
    new_elem.setAttribute("v", existing_queries.length);
    preferences.appendChild(new_elem);
    for (let i = 0; i < existing_queries.length; i++) {
      const q = existing_queries[i];
      if (q.name.length > 200)
        return callback(
          new Error("query name too long to be saved on osm.org")
        );
      const numParts = Math.ceil((q.query.length + q.name.length + 8) / 255);
      if (numParts > 9)
        return callback(new Error("query too long to be saved on osm.org"));
      let queryStr = `p=${numParts}`;
      queryStr += `&n=${q.name}`;
      queryStr += `&q=${q.query}`;
      // split into chunks of max 255 characters length
      queryStr = queryStr.match(/.{1,255}/g);

      for (let j = 0; j < numParts; j++) {
        const new_elem = dom.createElement("preference");
        new_elem.setAttribute(
          "k",
          `${configs.settingNamespace}_query_${i}_${j}`
        );
        new_elem.setAttribute("v", queryStr[j]);
        preferences.appendChild(new_elem);
      }
    }
    // upload to osm.org
    auth.xhr(
      {
        method: "PUT",
        path: "/api/0.6/user/preferences",
        headers: {"Content-Type": "text/xml"},
        content: dom.firstChild.outerHTML
      },
      (err) => {
        if (err) return callback(err);
        callback(null, existing_queries);
      }
    );
  });
}
