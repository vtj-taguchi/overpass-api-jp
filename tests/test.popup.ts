import {describe, expect, it} from "vitest";
import {featurePopupContent} from "../js/popup";

describe("featurePopupContent", () => {
  it("", () => {
    const feature = {
      type: "Feature",
      id: "node/270198479",
      properties: {
        type: "node",
        id: "270198479",
        tags: {
          amenity: "drinking_water",
          fountain: "roman_wolf",
          wikimedia_commons:
            "File:Roman Wolf Fountain, Villa Celimontana, Roma, Italia Sep 01, 2020 12-41-19 PM.jpeg"
        },
        relations: [],
        meta: {
          timestamp: "2022-10-03T09:17:16Z",
          version: "4",
          changeset: "126925539",
          user: "Friendly_Ghost",
          uid: "10875409"
        }
      },
      geometry: {
        type: "Point",
        coordinates: [12.4947844, 41.8837598]
      }
    };
    expect(featurePopupContent(feature)).toMatchSnapshot();
  });
  it("", () => {
    const feature = {
      type: "Feature",
      id: "relation/1243821",
      properties: {
        type: "relation",
        id: "1243821",
        tags: {
          architect: "André and Jean Polak",
          "architect:wikidata": "Q2848896",
          building: "yes",
          "contact:email": "info@atomium.be",
          "contact:phone": "+32 2 475 47 75",
          "contact:website": "https://www.atomium.be/",
          engineer: "André Waterkeyn",
          "engineer:wikidata": "Q523346",
          height: "102",
          image:
            "https://images.mapillary.com/dWS3qotev3rzfk3zwkB0nu/thumb-2048.jpg",
          layer: "1",
          mapillary: "291574785922785",
          name: "Atomium",
          start_date: "1958",
          "toilets:wheelchair": "yes",
          tourism: "attraction",
          type: "multipolygon",
          wheelchair: "limited",
          wikidata: "Q180901",
          wikipedia: "fr:Atomium"
        },
        relations: [],
        meta: {},
        geometry: "center"
      },
      geometry: {
        type: "Point",
        coordinates: [4.3415237, 50.894924]
      }
    };
    expect(featurePopupContent(feature)).toMatchSnapshot();
  });
});
