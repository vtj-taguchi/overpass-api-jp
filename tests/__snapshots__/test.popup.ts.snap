// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`featurePopupContent >  1`] = `"<h4 class=\\"title is-4\\"><span class=\\"t\\" data-t=\\"popup.node\\">Node</span> <a href=\\"//www.openstreetmap.org/node/270198479\\" target=\\"_blank\\">270198479</a> <a href=\\"//www.openstreetmap.org/edit?node=270198479\\" target=\\"_blank\\">✏</a></h4><h5 class=\\"subtitle is-5\\"><span class=\\"t\\" data-t=\\"popup.tags\\">Tags</span> <span class=\\"tag is-info is-light\\">3</span></h5><ul><li><span class=\\"is-family-monospace\\">amenity = drinking_water</span></li><li><span class=\\"is-family-monospace\\">fountain = roman_wolf</span></li><li><span class=\\"is-family-monospace\\">wikimedia_commons = <a href=\\"//commons.wikimedia.org/wiki/File:Roman Wolf Fountain, Villa Celimontana, Roma, Italia Sep 01, 2020 12-41-19 PM.jpeg\\" target=\\"_blank\\">File:Roman Wolf Fountain, Villa Celimontana, Roma, Italia Sep 01, 2020 12-41-19 PM.jpeg</a></span></li></ul><h4 class=\\"subtitle is-5\\"><span class=\\"t\\" data-t=\\"popup.metadata\\">Metadata</span></h4><ul><li><span class=\\"is-family-monospace\\">timestamp = 2022-10-03T09:17:16Z</span></li><li><span class=\\"is-family-monospace\\">version = 4</span></li><li><span class=\\"is-family-monospace\\">changeset = <a href=\\"//www.openstreetmap.org/changeset/126925539\\" target=\\"_blank\\">126925539</a></span></li><li><span class=\\"is-family-monospace\\">user = <a href=\\"//www.openstreetmap.org/user/Friendly_Ghost\\" target=\\"_blank\\">Friendly_Ghost</a></span></li><li><span class=\\"is-family-monospace\\">uid = 10875409</span></li></ul><h3 class=\\"subtitle is-5\\"><span class=\\"t\\" data-t=\\"popup.coordinates\\">Coordinates</span></h3><p><a href=\\"geo:41.8837598,12.4947844\\">41.8837598 / 12.4947844</a> <small>(lat/lon)</small></p>"`;

exports[`featurePopupContent >  2`] = `"<h4 class=\\"title is-4\\"><span class=\\"t\\" data-t=\\"popup.relation\\">Relation</span> <a href=\\"//www.openstreetmap.org/relation/1243821\\" target=\\"_blank\\">1243821</a> <a href=\\"//www.openstreetmap.org/edit?relation=1243821\\" target=\\"_blank\\">✏</a></h4><h5 class=\\"subtitle is-5\\"><span class=\\"t\\" data-t=\\"popup.tags\\">Tags</span> <span class=\\"tag is-info is-light\\">20</span></h5><ul><li><span class=\\"is-family-monospace\\">architect = André and Jean Polak</span></li><li><span class=\\"is-family-monospace\\">architect:wikidata = <a href=\\"//www.wikidata.org/wiki/Q2848896\\" target=\\"_blank\\">Q2848896</a></span></li><li><span class=\\"is-family-monospace\\">building = yes</span></li><li><span class=\\"is-family-monospace\\">contact:email = <a href=\\"mailto:info@atomium.be\\" target=\\"_blank\\">info@atomium.be</a></span></li><li><span class=\\"is-family-monospace\\">contact:phone = <a href=\\"tel:+32 2 475 47 75\\" target=\\"_blank\\">+32 2 475 47 75</a></span></li><li><span class=\\"is-family-monospace\\">contact:website = <a href=\\"https://www.atomium.be/\\" target=\\"_blank\\">https://www.atomium.be/</a></span></li><li><span class=\\"is-family-monospace\\">engineer = André Waterkeyn</span></li><li><span class=\\"is-family-monospace\\">engineer:wikidata = <a href=\\"//www.wikidata.org/wiki/Q523346\\" target=\\"_blank\\">Q523346</a></span></li><li><span class=\\"is-family-monospace\\">height = 102</span></li><li><span class=\\"is-family-monospace\\">image = <a href=\\"https://images.mapillary.com/dWS3qotev3rzfk3zwkB0nu/thumb-2048.jpg\\" target=\\"_blank\\">https://images.mapillary.com/dWS3qotev3rzfk3zwkB0nu/thumb-2048.jpg</a></span></li><li><span class=\\"is-family-monospace\\">layer = 1</span></li><li><span class=\\"is-family-monospace\\">mapillary = <a href=\\"https://www.mapillary.com/app?focus=photo&pKey=291574785922785\\" target=\\"_blank\\">291574785922785</a></span></li><li><span class=\\"is-family-monospace\\">name = Atomium</span></li><li><span class=\\"is-family-monospace\\">start_date = 1958</span></li><li><span class=\\"is-family-monospace\\">toilets:wheelchair = yes</span></li><li><span class=\\"is-family-monospace\\">tourism = attraction</span></li><li><span class=\\"is-family-monospace\\">type = multipolygon</span></li><li><span class=\\"is-family-monospace\\">wheelchair = limited</span></li><li><span class=\\"is-family-monospace\\">wikidata = <a href=\\"//www.wikidata.org/wiki/Q180901\\" target=\\"_blank\\">Q180901</a></span></li><li><span class=\\"is-family-monospace\\">wikipedia = <a href=\\"//fr.wikipedia.org/wiki/Atomium\\" target=\\"_blank\\">fr:Atomium</a></span></li></ul><h3 class=\\"subtitle is-5\\"><span class=\\"t\\" data-t=\\"popup.coordinates\\">Coordinates</span></h3><p><a href=\\"geo:50.894924,4.3415237\\">50.894924 / 4.3415237</a> <small>(lat/lon)</small></p>"`;
